core/src/main/java/jenkins/util/SystemProperties.java

```diff
@@ -60,40 +74,112 @@
+
     // declared in WEB-INF/web.xml
     public static final class Listener implements ServletContextListener, OnMaster {
 
-        /**
-         * The ServletContext to get the "init" parameters from.
-         */
-        @CheckForNull
-        private static ServletContext theContext;
-
         /**
          * Called by the servlet container to initialize the {@link ServletContext}.
          */
         @Override
-        @SuppressFBWarnings(value = "ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD",
-                justification = "Currently Jenkins instance may have one ond only one context")
         public void contextInitialized(ServletContextEvent event) {
-            theContext = event.getServletContext();
+            ServletContext theContext = event.getServletContext();
+            handler = new Handler() {
+                @Override
+                public String getString(String key) {
+                    if (StringUtils.isNotBlank(key)) {
+                        try {
+                            return theContext.getInitParameter(key);
+                        } catch (SecurityException ex) {
+                            // Log exception and go on
+                            LOGGER.log(Level.CONFIG, "Access to the property {0} is prohibited", key);
+                        }
+                    }
+                    return null;
+                }
+                @Override
+                public Map<String, String> getAllStrings() {
+                    Map<String, String> values = new HashMap<>();
+                    Enumeration<String> names = theContext.getInitParameterNames();
+                    while (names.hasMoreElements()) {
+                        String key = names.nextElement();
+                        String value = getString(key);
+                        if (value != null) {
+                            values.put(key, value);
+                        }
+                    }
+                    return values;
+                }
+            };
         }
 
         @Override
         public void contextDestroyed(ServletContextEvent event) {
-            theContext = null;
+            handler = NULL_HANDLER;
         }
 
     }
 
+    @Extension
+    public static final class SlaveCopier extends ComputerListener {
+        @Override
+        public void preOnline(Computer c, Channel channel, FilePath root, TaskListener listener) throws IOException, InterruptedException {
+            channel.call(new CopySystemProperties());
+        }
+        private static final class CopySystemProperties extends MasterToSlaveCallable<Void, RuntimeException> {
+            private static final long serialVersionUID = 1;
+            private final Map<String, String> allStrings;
+            @SuppressWarnings("unchecked")
+            CopySystemProperties() {
+                // Take a snapshot of all system properties and context variables available on the master at the time the agent starts.
```

Comment:
After discussion with @daniel-beck I think this should be limited to those system properties already used by `SystemProperties` on the master side.